### PR TITLE
libbpf-tools: Add support for IPv6 to tcprtt.

### PR DESCRIPTION
**Pull request by eiffel-fl**
Hi.

With this contribution, I added IPv6 support to `tcprtt`:

```
$ sudo ./tcprtt -b -c '::1'          
Tracing TCP RTT... Hit Ctrl-C to end.
^C
Local Address = ::1 
     usecs               : count    distribution
         0 -> 1          : 0        |                                        |
         2 -> 3          : 0        |                                        |
         4 -> 7          : 0        |                                        |
         8 -> 15         : 0        |                                        |
        16 -> 31         : 20       |***                                     |
        32 -> 63         : 266      |****************************************|
        64 -> 127        : 90       |*************                           |
```

If you see any ways to improve this contribution, feel free to share.

Best regards and thank you in advance.

### CODE REVIEW

File: libbpf-tools/tcprtt.bpf.c [hunk 1]
```diff
@@ -9,49 +9,119 @@
 #include "bits.bpf.h"
 #include "maps.bpf.h"
 
+/* Taken from kernel include/linux/socket.h. */
+#define AF_INET		2	/* Internet IP Protocol 	*/
```
Comment (yonghong-song) [inline libbpf-tools/tcprtt.bpf.c:13]: Just 'IP version 4' to make it easier to compare with 'IP version 6'.

File: libbpf-tools/tcprtt.bpf.c [hunk 2]
```diff
@@ -9,49 +9,119 @@
...
-		return 0;
-	if (targ_daddr && targ_daddr != BPF_CORE_READ(sk, __sk_common.skc_daddr))
-		return 0;
+
+	key.family = BPF_CORE_READ(sk, __sk_common.skc_family);
+	switch (key.family) {
+		case AF_INET:
+			/* If we set any of IPv6 address, we do not care about IPv4 ones. */
+			if (ipv6_is_not_zero(targ_saddr_v6) || ipv6_is_not_zero(targ_daddr_v6))
+				return 0;
```
Comment (yonghong-song) [inline libbpf-tools/tcprtt.bpf.c:82]: I think the above 'if' statement is not needed.

File: libbpf-tools/tcprtt.bpf.c [hunk 3]
```diff
@@ -9,49 +9,119 @@
...
 	if (targ_dport && targ_dport != BPF_CORE_READ(sk, __sk_common.skc_dport))
 		return 0;
-	if (targ_saddr && targ_saddr != BPF_CORE_READ(inet, inet_saddr))
-		return 0;
-	if (targ_daddr && targ_daddr != BPF_CORE_READ(sk, __sk_common.skc_daddr))
-		return 0;
+
+	key.family = BPF_CORE_READ(sk, __sk_common.skc_family);
+	switch (key.family) {
+		case AF_INET:
```
Comment (yonghong-song) [inline libbpf-tools/tcprtt.bpf.c:79]: pls align 'case' with 'switch' statement.

File: libbpf-tools/tcprtt.bpf.c [hunk 4]
```diff
@@ -9,49 +9,119 @@
...
+				return 0;
+
+		if (targ_saddr && targ_saddr != BPF_CORE_READ(inet, inet_saddr))
+			return 0;
+
+		if (targ_daddr && targ_daddr != BPF_CORE_READ(sk, __sk_common.skc_daddr))
+			return 0;
+
+		break;
+		case AF_INET6:
```
Comment (yonghong-song) [inline libbpf-tools/tcprtt.bpf.c:91]: same here.

File: libbpf-tools/tcprtt.bpf.c [hunk 5]
```diff
@@ -9,49 +9,119 @@
...
+			/*
+			 * Reciprocal of the above: if we set any of IPv4 address, we do not care
+			 * about IPv6 ones.
+			 */
+			if (targ_saddr || targ_daddr)
+				return 0;
```
Comment (yonghong-song) [inline libbpf-tools/tcprtt.bpf.c:97]: the above is not needed.

File: libbpf-tools/tcprtt.bpf.c [hunk 6]
```diff
@@ -9,49 +9,119 @@
...
+		if (ipv6_is_not_zero(targ_saddr_v6)
+			&& ipv6_are_different(targ_saddr_v6, BPF_CORE_READ(inet, pinet6, saddr.in6_u.u6_addr8)))
+			return 0;
+
+		if (ipv6_is_not_zero(targ_daddr_v6)
+			&& ipv6_are_different(targ_daddr_v6, BPF_CORE_READ(sk, __sk_common.skc_v6_daddr.in6_u.u6_addr8)))
+			return 0;
+
+		break;
+		default:
```
Comment (yonghong-song) [inline libbpf-tools/tcprtt.bpf.c:108]: same here.

File: libbpf-tools/tcprtt.bpf.c [hunk 7]
```diff
@@ -9,49 +9,119 @@
...
+			return 0;
```
Comment (yonghong-song) [inline libbpf-tools/tcprtt.bpf.c:109]: align here as well.

File: libbpf-tools/tcprtt.bpf.c [hunk 8]
```diff
@@ -9,49 +9,119 @@
...
+		bpf_probe_read_kernel(key.addr, sizeof(inet->inet_saddr), &inet->inet_saddr);
 	else if (targ_raddr_hist)
-		key = BPF_CORE_READ(inet, sk.__sk_common.skc_daddr);
+		if (key.family == AF_INET6)
+			bpf_probe_read_kernel(&key.addr, sizeof(key.addr), BPF_CORE_READ(sk, __sk_common.skc_v6_daddr.in6_u.u6_addr8));
 	else
-		key = 0;
+		bpf_probe_read_kernel(&key.addr, sizeof(inet->sk.__sk_common.skc_daddr), &inet->sk.__sk_common.skc_daddr);
+	else
+		key.family = 0;
```
Comment (yonghong-song) [inline libbpf-tools/tcprtt.bpf.c:123]: The whole above if/else/else if is really confusing. Please format with proper indentation and with '{ }' properly.

File: libbpf-tools/tcprtt.c [hunk 9]
```diff
@@ -169,39 +188,59 @@ static void sig_handler(int sig)
...
 		else
 			printf("All Addresses = ****** ");
+
+		if (env.laddr_hist || env.raddr_hist) {
+			__u16 family = next_key.family;
+			char str[INET6_ADDRSTRLEN];
+
+			if (family != AF_INET && family != AF_INET6) {
+				fprintf(stderr, "expected AF_INET (%d) or AF_INET6 (%d) for family, got: %d\n", AF_INET, AF_INET6, family);
+
```
Comment (yonghong-song) [inline libbpf-tools/tcprtt.c:216]: extra line not necessary.

File: libbpf-tools/tcprtt.c [hunk 10]
```diff
@@ -169,39 +188,59 @@ static void sig_handler(int sig)
...
+				return -1;
+			}
+
+			if (!inet_ntop(family, next_key.addr, str, sizeof str)) {
+				perror("converting IP to string:");
+
```
Comment (yonghong-song) [inline libbpf-tools/tcprtt.c:222]: same here.

File: libbpf-tools/tcprtt.c [hunk 11]
```diff
@@ -225,6 +265,14 @@ int main(int argc, char **argv)
 	if (err)
 		return err;
 
+	if ((env.laddr || env.raddr)
+		&& (!memcmp(env.laddr_v6, zero_addr_v6, sizeof env.laddr_v6) || !memcmp(env.raddr_v6, zero_addr_v6, sizeof env.raddr_v6)))
+	{
+		fprintf(stderr, "It is not permitted to filter by both IPv6 and IPv4");
+
```
Comment (yonghong-song) [inline libbpf-tools/tcprtt.c:272]: same here.

**Review by yonghong-song — COMMENTED** [review]
In general, this is a good addition to tcprtt tool. Some formatting and cleanup is needed though.

File: libbpf-tools/tcprtt.c [hunk 12]
```diff
@@ -225,6 +263,12 @@ int main(int argc, char **argv)
 	if (err)
 		return err;
 
+	if ((env.laddr || env.raddr)
+		&& (!memcmp(env.laddr_v6, zero_addr_v6, sizeof env.laddr_v6) || !memcmp(env.raddr_v6, zero_addr_v6, sizeof env.raddr_v6))) {
+		fprintf(stderr, "It is not permitted to filter by both IPv6 and IPv4");
```
Comment (yonghong-song) [inline libbpf-tools/tcprtt.c:268]: Please add '\n' as the end of error string. Maybe other places need '\n' as well. Please double check.

File: libbpf-tools/tcprtt.bpf.c [hunk 13]
```diff
@@ -9,49 +9,120 @@
...
+
+		if (targ_daddr && targ_daddr != BPF_CORE_READ(sk, __sk_common.skc_daddr))
+			return 0;
+
+		break;
+	case AF_INET6:
+		/*
+			* Reciprocal of the above: if we set any of IPv4 address, we do not care
+			* about IPv6 ones.
+			*/
```
Comment (yonghong-song) [inline libbpf-tools/tcprtt.bpf.c:95]: alignment issue in the above.

File: libbpf-tools/tcprtt.c [hunk 14]
```diff
@@ -225,6 +263,12 @@ int main(int argc, char **argv)
 	if (err)
 		return err;
 
+	if ((env.laddr || env.raddr)
+		&& (!memcmp(env.laddr_v6, zero_addr_v6, sizeof env.laddr_v6) || !memcmp(env.raddr_v6, zero_addr_v6, sizeof env.raddr_v6))) {
```
Comment (yonghong-song) [inline libbpf-tools/tcprtt.c:267]: Did you check when with '-a'/'-A'? if env.laddr/env.raddr is not NULL, and env.laddr_v2/raddr_v6 is NULL, the below error is triggered.
Let us use the format 'sizeof(...)' which is more uniform compared to other usage in libbpf-tools.

File: libbpf-tools/tcprtt.c (hunk 12 above)
Comment (yonghong-song) [inline libbpf-tools/tcprtt.c:268]: Please add a '\n' at the end of error message.

File: libbpf-tools/tcprtt.c (hunk 14 above)
Comment (yonghong-song) [inline libbpf-tools/tcprtt.c:267]: This is not correct. The '!' is not needed here for correctness.

File: libbpf-tools/tcprtt.c [hunk 15]
```diff
@@ -139,6 +144,20 @@ static error_t parse_arg(int key, char *arg, struct argp_state *state)
 		}
 		env.raddr = addr.s_addr;
 		break;
+	case 'c':
+		if (inet_pton(AF_INET6, arg, &addr_v6) < 1) {
+			fprintf(stderr, "invalid local IPv6 address: %s\n", arg);
+			argp_usage(state);
+		}
+		memcpy(env.laddr_v6, &addr_v6, sizeof env.laddr_v6);
```
Comment (yonghong-song) [inline libbpf-tools/tcprtt.c:152]: please use `sizeof(env.laddr_v6)`. The same for other instances in this file. Just do a search for 'sizeof' and make changes to all those not 'sizeof(...)' cases.

File: libbpf-tools/tcprtt.bpf.c [hunk 16]
```diff
@@ -9,49 +9,120 @@
...
+
+
 const volatile bool targ_laddr_hist = false;
 const volatile bool targ_raddr_hist = false;
 const volatile bool targ_show_ext = false;
 const volatile __u16 targ_sport = 0;
 const volatile __u16 targ_dport = 0;
 const volatile __u32 targ_saddr = 0;
 const volatile __u32 targ_daddr = 0;
+const volatile __u8 targ_saddr_v6[IPV6_LEN] = { 0 };
```
Comment (chenhengqi) [inline libbpf-tools/tcprtt.bpf.c:24]: Just use `= {};`, here and all other occurrences.

File: libbpf-tools/tcprtt.bpf.c [hunk 17]
```diff
@@ -9,49 +9,120 @@
...
 	__type(value, struct hist);
 } hists SEC(".maps");
 
 static struct hist zero;
 
+/*
+ * We cannot use the following:
+ * __builtin_memcmp(targ_*addr_v6, *, sizeof(targ_*addr_v6));
+ * Indeed, by using the builtin, we would discard the volatile qualifier of
+ * targ_*addr_v6, so the compiler would optimize it and replaces the call
```
Comment (chenhengqi) [inline libbpf-tools/tcprtt.bpf.c:43]: I don't understand the optimization mentioned here.

Comment (yonghong-song) [reply libbpf-tools/tcprtt.bpf.c:43]: @chenhengqi This is related to quirks in llvm compiler which I have not investigated it yet and I don't know whether upstream has any intention to change or not. Basically you have something like
```
const volatile __u8 targ_saddr_v6[IPV6_LEN] = {};
```

If we directly use targ_saddr_v6 in bpf program as most use cases in kernel bpf selftests, targ_saddr_v6 will not be treated as a 'const' and that is why its value can be changed after map is created but before program is loaded with libbpf skeleton.

But there is an abnormal here. If 'targ_saddr_v6' is used as a parameter to `__builtin_memcpy`, then the content of `target_saddr_v6` is treated as 'const' so the whole __builtin_memcpy(targ_saddr_v6, zero_addr_v6) is optimized away. 

I am aware of this issue and have not got chance to clarify this in upstream. I will check gcc behavior for this case as well and then will ask upstream.

Comment (yonghong-song) [reply libbpf-tools/tcprtt.bpf.c:43]: I filed an issue in llvm-project issue site: https://github.com/llvm/llvm-project/issues/63792
Feel to keep track of it or add a comment to it if you have any.

**Pull request merged**